tf-psa-crypto/scripts/config.py

```diff
@@ -0,0 +1,207 @@
+    an implementation of the relevant functions and an xxx_alt.h header.
+    """
+    if name in (
+            'MBEDTLS_PLATFORM_GMTIME_R_ALT',
+            'MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT',
+            'MBEDTLS_PLATFORM_MS_TIME_ALT',
+            'MBEDTLS_PLATFORM_ZEROIZE_ALT',
+    ):
+        # Similar to non-platform xxx_ALT, requires platform_alt.h
+        return False
+    return name.startswith('MBEDTLS_PLATFORM_')
+
+def include_in_full(name):
+    """Rules for symbols in the "full" configuration."""
+    if name in EXCLUDE_FROM_FULL:
+        return False
+    if name.endswith('_ALT'):
+        return is_seamless_alt(name)
+    return True
+
+def full_adapter(name, value, active):
+    """Config adapter for "full"."""
+    if not is_boolean_setting(name, value):
+        return active
+    return include_in_full(name)
+
+
+class TFPSACryptoConfigFile(config_common.ConfigFile):
+    """Representation of a TF PSA Crypto configuration file."""
+
+    _path_in_tree = 'include/psa/crypto_config.h'
+    default_path = [_path_in_tree,
+                    os.path.join(os.path.dirname(__file__),
+                                 os.pardir,
+                                 _path_in_tree),
+                    os.path.join(os.path.dirname(os.path.abspath(os.path.dirname(__file__))),
+                                 _path_in_tree)]
+
+    def __init__(self, filename=None):
+        super().__init__(self.default_path, 'TF-PSA-Crypto', filename)
+
+
+class TFPSACryptoConfig(config_common.Config):
+    """Representation of the TF PSA Crypto configuration.
+
+    See the documentation of the `Config` class for methods to query
+    and modify the configuration.
+    """
+
+    def __init__(self, filename=None):
+        """Read the PSA crypto configuration files."""
+
+        super().__init__()
+        configfile = TFPSACryptoConfigFile(filename)
+        self.configfiles.append(configfile)
+        self.settings.update({name: config_common.Setting(configfile, active, name, value, section)
+                             for (active, name, value, section) in configfile.parse_file()})
+
+    def set(self, name, value=None):
+        """Set name to the given value and make it active."""
+
+        if name in PSA_UNSUPPORTED_FEATURE:
+            raise ValueError(f'Feature is unsupported: \'{name}\'')
+        if name in PSA_UNSTABLE_FEATURE:
+            raise ValueError(f'Feature is unstable: \'{name}\'')
+
+        if name not in self.settings:
+            self._get_configfile().templates.append((name, '', f'#define {name} '))
+
+        # Default value for PSA macros is '1'
+        if name.startswith('PSA_') and not value:
+            value = '1'
+
+        super().set(name, value)
+
+
+class TFPSACryptoConfigTool(config_common.ConfigTool):
+    """Command line TF PSA Crypto config file manipulation tool."""
+
+    def __init__(self):
+        super().__init__(TFPSACryptoConfigFile.default_path[0])
```

Comment:
I am a bit confused as to why we initiate a full list of default paths above and then only use the `_path_in_tree = 'include/psa/crypto_config.h'`

Wouldn't that work?

`super().__init__(TFPSACryptoConfigFile._path_in_tree)`

Reply:
This parameter is only used for the help menu and will generate a lot of 'noise' in the menu:
```
options:
  -h, --help            show this help message and exit
  --file FILE, -f FILE  File to read (and modify if requested). Default:
                        ['include/psa/crypto_config.h',
                        '/home/gabmez01/Work/mbedTLS/9157/mbedtls/tf-psa-
                        crypto/scripts/../include/psa/crypto_config.h',
                        '/home/gabmez01/Work/mbedTLS/9157/mbedtls/tf-psa-
                        crypto/include/psa/crypto_config.h'].
  --force, -o           For the set command, if SYMBOL is not present, add a
                        definition for it.
  --write FILE, -w FILE
                        File to write to instead of the input file.

```

Reply:
> I am a bit confused as to why we initiate a full list of default paths above and then only use the `_path_in_tree = 'include/psa/crypto_config.h'`

This is not the only usage:
https://github.com/Mbed-TLS/mbedtls/blob/1a7bbe1ca8d9bdb344cb55a06b5d303df1559281/tf-psa-crypto/scripts/config.py#L151

> Wouldn't that work?
> 
> `super().__init__(TFPSACryptoConfigFile._path_in_tree)`

The variables starts with `_` are considered private in python. The `_path_in_tree` should not be used outside of its class.